src/main/groovy/com/bmuschko/gradle/docker/DockerJavaApplication.groovy

```diff
@@ -59,42 +55,48 @@ class DockerJavaApplication {
      * Helper Instruction to allow customizing generated ENTRYPOINT/CMD.
      */
     static class CompositeExecInstruction implements Dockerfile.Instruction {
-        private final List<Dockerfile.Instruction> instructions = new ArrayList<Dockerfile.Instruction>()
+        private final ListProperty<Dockerfile.Instruction> instructions
+
+        CompositeExecInstruction(Project project) {
+            instructions = project.objects.listProperty(Dockerfile.Instruction)
+        }
 
         @Nested
-        List<Dockerfile.Instruction> getInstructions() {
+        ListProperty<Dockerfile.Instruction> getInstructions() {
             instructions
         }
 
         @Override
         String getKeyword() { '' }
 
-        @Override
-        String getText() {
-            build()
+        void clear() {
+            instructions.set([])
         }
 
         @Override
-        String build() { instructions*.getText().join(System.getProperty('line.separator')) }
-
-        void clear() {
-            instructions.clear()
+        String getText() {
+            List<Dockerfile.Instruction> viableInstructions = instructions.get().findAll { it.text }
+            viableInstructions.collect { it.getText() }.join(System.getProperty('line.separator'))
         }
 
         void defaultCommand(String... command) {
-            instructions << new Dockerfile.DefaultCommandInstruction(command)
+            addInstruction(new Dockerfile.DefaultCommandInstruction(command))
         }
 
-        void defaultCommand(Closure command) {
-            instructions << new Dockerfile.DefaultCommandInstruction(command)
+        void defaultCommand(Provider<List<String>> commandProvider) {
+            addInstruction(new Dockerfile.DefaultCommandInstruction(commandProvider))
         }
 
         void entryPoint(String... entryPoint) {
-            instructions << new Dockerfile.EntryPointInstruction(entryPoint)
+            addInstruction(new Dockerfile.EntryPointInstruction(entryPoint))
+        }
+
+        void entryPoint(Provider<List<String>> entryPointProvider) {
```

Comment:
Don't you want this to be a `ListProvider<String>` instead?

Reply:
If you look at the API of `ListProperty` you will see that the class extends from `Provider<List<T>>`. Passing in a `ListProperty` should work here. I don't necessarily want to expose a `Property` as the provided value should be immutable. Therefore, a `Provider` is likely the better choice.

Reply:
That makes sense to me.